Hyperion/SerializerFactories/EnumerableSerializerFactory.cs

```diff
@@ -68,33 +75,84 @@ private static Type GetEnumerableType(Type type)
+        {
+            var arg = Expression.Parameter(typeof(object));
+            var castArg = Expression.Convert(arg, argType);
+            var call = Expression.New(ctor, new Expression[] { castArg });
+            var castRes = Expression.Convert(call, typeof(object));
+            var lambda = Expression.Lambda<Func<object, object>>(castRes, arg);
+            var compiled = lambda.Compile();
+            return compiled;
+        }
+
+        private static Action<object, object> CompileMethodToDelegate(MethodInfo method, Type instanceType, Type argType)
+        {
+            var instance = Expression.Parameter(typeof(object));
+            var arg = Expression.Parameter(typeof(object));
+            var castInstance = Expression.Convert(instance, instanceType);
+            var castArg = Expression.Convert(arg, argType);
+            var call = Expression.Call(castInstance, method, new Expression[] { castArg });
+            var lambda = Expression.Lambda<Action<object, object>>(call, instance, arg);
+            var compiled = lambda.Compile();
+            return compiled;
+        }
+
         public override ValueSerializer BuildSerializer(Serializer serializer, Type type,
             ConcurrentDictionary<Type, ValueSerializer> typeMapping)
         {
             var x = new ObjectSerializer(type);
             typeMapping.TryAdd(type, x);
+
             var preserveObjectReferences = serializer.Options.PreserveObjectReferences;
 
             var elementType = GetEnumerableType(type) ?? typeof(object);
             var elementSerializer = serializer.GetSerializerByType(elementType);
 
             var countProperty = type.GetTypeInfo().GetProperty("Count");
-            var addRange = type.GetTypeInfo().GetMethod("AddRange");
-            var add = type.GetTypeInfo().GetMethod("Add");
+            var constructor = GetEnumerableConstructor(type);
+            var construct = constructor != null 
+                ? CompileCtorToDelegate(constructor, elementType.MakeArrayType()) 
+                : null;
+            var addRangeMethod = type.GetTypeInfo().GetMethod("AddRange");
+            var addRange = construct == null && addRangeMethod != null
+                 ? CompileMethodToDelegate(addRangeMethod, type, elementType.MakeArrayType())
+                 : null;
+            var addMethod = type.GetTypeInfo().GetMethod("Add");
+            var add = construct == null && addRange == null && addMethod != null
+                ? CompileMethodToDelegate(addMethod, type, elementType)
+                : null;
 
             Func<object, int> countGetter = o => (int)countProperty.GetValue(o);
-
-
+            
             ObjectReader reader = (stream, session) =>
             {
-                var instance = Activator.CreateInstance(type);
-                if (preserveObjectReferences)
+                var count = stream.ReadInt32(session);
+                if (construct != null)
```

Comment:
You can move those null checks outside of the `ObjectReader` implementation - it's not like the constructor/addRange/add presence for the same type will change from one deserializer call to another. Instead you can make those checks outside and return one of the 3 different `ObjectReader` implementations (depending if constructor/addRange/add is available).